modules/authtoservices.cpp

```diff
@@ -0,0 +1,499 @@
+        HandleHelpCommand(sLine);
+
+        CTable Mechanisms;
+        Mechanisms.AddColumn(t_s("Mechanism"));
+        Mechanisms.AddColumn(t_s("Description"));
+
+        for (const auto& it : SupportedMechanisms) {
+            Mechanisms.AddRow();
+            Mechanisms.SetCell(t_s("Mechanism"), it.szName);
+            Mechanisms.SetCell(t_s("Description"), it.sDescription.Resolve());
+        }
+
+        PutModule(t_s("The following mechanisms are available:"));
+        PutModule(Mechanisms);
+    }
+
+    ~CAuthToServices() override {}
+
+    bool LoadOtherModuleNV(const CString &otherModule, MCString& result) {
+        CString sPath = GetNetwork()->GetNetworkPath() + "/moddata/"
+                        + otherModule + "/.registry";
+        return result.ReadFromDisk(sPath) == MCString::MCS_SUCCESS;
+    }
+
+    bool OnLoad(const CString& sArgs, CString& sMessage) override {
+        if (sArgs == "as_sasl") {
+            if (LoadOtherModuleNV("sasl", m_mssSaslRegistry)) {
+                SetNV("RequireSASL", m_mssSaslRegistry["require_auth"]);
+                SetNV("SASLMechanisms", m_mssSaslRegistry["mechanisms"]);
+                SetNV("Account", m_mssSaslRegistry["username"]);
+                SetNV("Password", m_mssSaslRegistry["password"]);
+
+                PutModule(t_s("Migrated configuration from SASL module."));
+            } else {
+                PutModule(t_s("Could not migrate configuration from SASL module."));
+            }
+        }
+        else if (sArgs == "as_nickserv") {
+            if (LoadOtherModuleNV("nickserv", m_mssNickservRegistry)) {
+                SetNV("Password", m_mssNickservRegistry["Password"]);
+                SetNV("IdentifyCmd", m_mssNickservRegistry["IdentifyCmd"]);
+
+                if(!m_mssSaslRegistry["NickServName"].empty()) {
+                    SetNV("NickServName", m_mssSaslRegistry["NickServName"]);
+                }
+
+                PutModule(t_s("Migrated configuration from NickServ module."));
+            } else {
+                PutModule(t_s("Could not migrate configuration from NickServ module."));
+            }
+        }
+
+        SetNV("NotFirstLoad", "true");
+
+        if (GetNV("IdentifyCmd").empty()) {
+            SetNV("IdentifyCmd", "NICKSERV IDENTIFY {account} {password}");
+        }
+
+        return true;
+    }
+
+    const CString GetAccount() {
+        if(GetNV("Account").empty()) {
+            CString sNick = GetNetwork()->GetNick();
+            DEBUG("Account is empty, using " + sNick + " instead");
+            return sNick;
+        }
+        return GetNV("Account");
+    }
+
+    // NickServ authentication
+
+    void NickServAuthenticate() {
+        MCString msValues;
+        msValues["account"] = GetAccount();
+        msValues["password"] = GetNV("Password");
+        PutIRC(CString::NamedFormat(GetNV("IdentifyCmd"), msValues));
+    }
+
+    void HandleMessage(CNick& Nick, const CString& sMessage) {
```

Comment:
Need a way to disable this, e.g. via additional "mechanism"

Reply:
:+1:, this can also be a security flaw on sasl only networks where there isn't a NickServ nick. It might be possible for a user to get sasl passwords for ZNC users (and other clients that have similar logic) because they can change their nick to `NickServ` and send one of messages below to get clients to send passwords to them. Some servers can offer blocking nick names out from use like NickServ and common typos, but that is relying on the server being configured correctly.

I think the spirit of what we're trying to do with #1494 / #1230 is to encourage users to use sasl over nickserv (instead of, not with) because it has many benefits, including the decrease in risks where passwords can be erroneously taken. If its not by automated tooling, even there are known impersonation attacks in the wild with bots using nicknames similar to nickserv such as nicksarv and others to trick users into sending them their passwords.

If a network offers SASL I don't think users should consider using nickserv to idenify. Even if the sasl capability is currently missing or unavailable during connection, there can be room for more sophisticated attacks where you can abuse this feature of clients while service may be offline or down for maintenance. I'd avoid the room for downgrade attacks from sasl to nickserv auth. The original sasl module also has `requireauth` option to disconnect if we can't authenticate before registration (providing upstream server supports capabilities).

---

The problem right now is that some users do not know what sasl is, and they are already familiar with nickserv and thus would setup the nickserv module instead. I would suggest that we go with an approach that allows the module to be configured to only allow sasl authentication, and if at any time the module determines it can "upgrade" from from nickserv to sasl auth, it should then set this flag so it would never downgrade without user consent. Or even as alternative to merging the modules together, make the current NickServ module fail if it detects the server supports sasl and facilitate the user on switching to the sasl module instead (and some way to continue with nickserv if the user really knows what they are doing and understands the risks). However this approach may be a bit more user hostile than people would like.